File: Algorithms/Python/17 Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        def backtracking(i):
            nonlocal tmp
            if i == len(digits):
                ans.append(''.join(tmp))
                return
            
            for alpha in num_to_alpha[int(digits[i])]:
                tmp[i] = alpha
                backtracking(i + 1)

        if len(digits) == 0:
            return []
        
        ans = []
        tmp = ['a'] * len(digits)
        num_to_alpha = {2:'abc', 3:'def', 4:'ghi', 5:'jkl', 6:'mno', 7:'pqrs', 8:'tuv', 9:'wxyz'}
        
        backtracking(0)

        return ans

File: Algorithms/CPP/17 Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) {
            return vector<string>();    
        }

        tmp = string(digits.size(), 'a');
        this->digits = digits;
        backtracking(0);
        
        return ans;
    }

    void backtracking(int i) {
        if (i == digits.size()) {
            ans.push_back(tmp);
            return;
        }

        for (const auto& alpha : number_to_alpha[digits[i]]) {
            tmp[i] = alpha;
            backtracking(i + 1);
        }
    }

    Solution() {
        number_to_alpha['2'] = "abc";
        number_to_alpha['3'] = "def";
        number_to_alpha['4'] = "ghi";
        number_to_alpha['5'] = "jkl";
        number_to_alpha['6'] = "mno";
        number_to_alpha['7'] = "pqrs";
        number_to_alpha['8'] = "tuv";
        number_to_alpha['9'] = "wxyz";
    }

private:
    map<char, string> number_to_alpha;
    string digits;
    string tmp;
    vector<string> ans;
};
